scripts/backend_tests.py

```diff
@@ -211,7 +211,24 @@ def _get_all_test_targets(test_path=None, include_load_tests=True):
     """
     def _convert_to_test_target(path):
         """Remove the .py suffix and replace all slashes with periods."""
-        return os.path.relpath(path, os.getcwd())[:-3].replace('/', '.')
+        classes = []
+        with open(path, 'r') as f:
+            lines = f.readlines()
+            for line in lines:
+                if line.startswith('class'):
+                    start_index = line.find(' ') + 1
+                    end_index = line.find('Tests(')
```

Comment:
I think this is fragile. It just takes someone to not name their class properly -- and then the test won't run, but no one will notice it.

Instead, could you do a check based on what the class is a subclass of? You might be able to use stuff like inspect.getmro (grep for that in the codebase) to do this more semantically, but I think the correct definition of test classes is "is a subclass of the base test class", not "is named XXX".

Reply:
I actually tried using importlib to import module from the test_target_path but I get errors there as follows:

```
Checking if node.js is installed in /home/ankita/Desktop/opensrc/oppia_tools
Checking whether Google App Engine is installed in /home/ankita/Desktop/opensrc/oppia_tools/google_appengine_1.9.67/google_appengine
Checking whether google-cloud-sdk is installed in /home/ankita/Desktop/opensrc/oppia_tools/google-cloud-sdk-251.0.0/google-cloud-sdk
Traceback (most recent call last):
  File "scripts/backend_tests.py", line 395, in <module>
    main()
  File "scripts/backend_tests.py", line 290, in main
    include_load_tests=include_load_tests)
  File "scripts/backend_tests.py", line 246, in _get_all_test_targets
    _convert_to_test_target(os.path.join(base_path, root)))
  File "scripts/backend_tests.py", line 219, in _convert_to_test_target
    python_module = importlib.import_module(test_target_path)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/ankita/Desktop/opensrc/oppia/utils_test.py", line 23, in <module>
    from core.tests import test_utils
  File "/home/ankita/Desktop/opensrc/oppia/core/tests/test_utils.py", line 30, in <module>
    from core.domain import collection_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/collection_services.py", line 33, in <module>
    from core.domain import exp_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/exp_services.py", line 38, in <module>
    from core.domain import email_subscription_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/email_subscription_services.py", line 19, in <module>
    from core.domain import email_manager
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/email_manager.py", line 22, in <module>
    from core.domain import config_domain
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/config_domain.py", line 22, in <module>
    import schema_utils
  File "/home/ankita/Desktop/opensrc/oppia/schema_utils.py", line 32, in <module>
    from core.domain import html_cleaner  # pylint: disable=relative-import
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/html_cleaner.py", line 24, in <module>
    import bleach
ImportError: No module named bleach
```

Is there some path issues and do I need to insert path for the modules?

Reply:
Possibly -- see core/tests/gae_suite.py.

But if this becomes too tricky/messy, then perhaps a lexical analysis of the "Abc" in `class XYZ(Abc)` might be fine -- maybe ensure it is from a given set of strings. That's more fragile though, so better to avoid if possible, but it might be a good middle ground if it proves too hard to use getmro().

Reply:
I tried adding the paths as in tests/gae_suite.py but I still get errors for modules being imported.

Also, for lexical analysis, the classes should inherit from a fixed class but in many files, we have new base classes and then test classes as a subclass of them. Would it be not better if we could have a test to ensure that all test classes should end in `Test/Tests`?

Reply:
> Also, for lexical analysis, the classes should inherit from a fixed class but in many files, we have new base classes and then test classes as a subclass of them. Would it be not better if we could have a test to ensure that all test class names should end in Test/Tests?

That could also work, but how would you do that? Such a test would still need to identify which classes are "test classes".

And yeah, that is the drawback of lexical analysis, we'd probably have to maintain a list of "allowed subclasses" or something, but that is indeed icky.

Reply:
Maybe we should use [metaclasses](https://avacariu.me/articles/2016/using-a-metaclass-to-keep-track-of-subclasses) to keep track of subclasses?

Reply:
This is actually fixed now. I added paths similar to that in `core/tests/gae_suite.py`. I missed pylint path earlier since it was not present in `gae_suite.py` but test files like `pylint_extensions_tests` require pylint import as well and so adding that to path fixed the issue!